File: src/js/core/src/Resolver.ts
import { Options } from './Options.js';
import { Prng } from './Prng.js';
import { Color } from './Utils/Color.js';
import { CircularColorReferenceError } from './Error/CircularColorReferenceError.js';
import type { Style } from './Style.js';
import type { Component } from './Style/Component.js';
import type {
  StyleOptionsFlipValue,
  StyleOptionsColorFillValue,
  StyleOptions,
} from './StyleOptions.js';

/**
 * Bundles the three inputs needed to derive any deterministic value for an
 * avatar — the {@link Style}, the validated user {@link Options}, and a
 * seeded {@link Prng} — and exposes them as named accessors. Each accessor
 * memoizes its result so that repeated calls cannot drift, and the memo
 * doubles as the snapshot returned by {@link resolved}.
 */
export class Resolver<D = unknown> {
  #style: Style<D>;
  #options: Options<D>;
  #prng: Prng;
  #colorResolving: string[] = [];
  #result: Record<string, unknown> = {};

  constructor(style: Style<D>, options: Options<D>) {
    this.#style = style;
    this.#options = options;
    this.#prng = new Prng(this.seed());
  }

  seed(): string {
    return this.#memo('seed', () => this.#options.seed() ?? '');
  }

  size(): number | undefined {
    return this.#memo('size', () => this.#options.size());
  }

  idRandomization(): boolean {
    return this.#memo('idRandomization', () => this.#options.idRandomization() ?? false);
  }

  title(): string | undefined {
    return this.#memo('title', () => this.#options.title());
  }

  flip(): StyleOptionsFlipValue {
    return this.#memo(
      'flip',
      () => this.#prng.pick('flip', this.#options.flip()) ?? 'none',
    );
  }

  fontFamily(): string {
    return this.#memo(
      'fontFamily',
      () => this.#prng.pick('fontFamily', this.#options.fontFamily()) ?? 'system-ui',
    );
  }

  fontWeight(): number {
    return this.#memo(
      'fontWeight',
      () => this.#prng.pick('fontWeight', this.#options.fontWeight()) ?? 400,
    );
  }

  scale(): number {
    return this.#memo(
      'scale',
      () => this.#prng.float('scale', this.#options.scale()) ?? 1,
    );
  }

  borderRadius(): number {
    return this.#memo(
      'borderRadius',
      () => this.#prng.float('borderRadius', this.#options.borderRadius()) ?? 0,
    );
  }

  rotate(): number {
    return this.#memo(
      'rotate',
      () => this.#prng.float('rotate', this.#options.rotate()) ?? 0,
    );
  }

  translateX(): number {
    return this.#memo(
      'translateX',
      () => this.#prng.float('translateX', this.#options.translateX()) ?? 0,
    );
  }

  translateY(): number {
    return this.#memo(
      'translateY',
      () => this.#prng.float('translateY', this.#options.translateY()) ?? 0,
    );
  }

  /**
   * Selects a variant for the given component. Depending on what was passed
   * as `${name}Variant` in the input data:
   *
   * - `undefined`: PRNG picks from all style variants using their weights.
   * - `string` or `string[]`: PRNG picks from the given subset (weight 1 each).
   * - `Record<string, number>`: PRNG picks using the provided weights.
   *
   * Only variants that exist in the style definition are considered.
   */
  variant(name: string): string | undefined {
    return this.#memo(`${name}Variant`, () => {
      const component = this.#style.components().get(name);

      if (!component || !this.#isVisible(name, component)) {
        return undefined;
      }

      const raw = this.#options.componentVariant(component.sourceName());
      const variants = component.variants();
      let entries: [string, number][];

      if (raw === undefined) {
        entries = Array.from(variants).map(([v, variant]) => [
          v,
          variant.weight(),
        ]);
      } else if (Array.isArray(raw)) {
        entries = raw
          .filter((v): v is string => variants.has(v))
          .map((v) => [v, 1]);
      } else {
        entries = Object.entries(raw).filter(([v]) => variants.has(v));
      }

      return this.#prng.weightedPick(`${name}Variant`, entries);
    });
  }

  color(name: string): readonly string[] {
    return this.#memo(`${name}Color`, () => this.#resolveColor(name));
  }

  colorFill(name: string): StyleOptionsColorFillValue {
    return this.#memo(
      `${name}ColorFill`,
      () =>
        this.#prng.pick(`${name}ColorFill`, this.#options.colorFill(name)) ??
        'solid',
    );
  }

  colorAngle(name: string): number {
    return this.#memo(
      `${name}ColorAngle`,
      () => this.#prng.float(`${name}ColorAngle`, this.#options.colorAngle(name)) ?? 0,
    );
  }

  /**
   * Per-component transform values are render-time decorations derived per
   * `<use>` reference, not user-input options that should round-trip. They
   * are intentionally not memoized and so never appear in {@link resolved}.
   */
  componentTransform(name: string): {
    rotate: number;
    translateX: number;
    translateY: number;
    scale: number;
  } {
    const component = this.#style.components().get(name);

    return {
      rotate: this.#pickFloat(name, 'Rotate', component?.rotate() ?? [], 0),
      translateX: this.#pickFloat(name, 'TranslateX', component?.translate().x() ?? [], 0),
      translateY: this.#pickFloat(name, 'TranslateY', component?.translate().y() ?? [], 0),
      scale: this.#pickFloat(name, 'Scale', component?.scale() ?? [], 1),
    };
  }

  /**
   * Returns every value that has been touched during this resolution. Only
   * memoized entries are included; unset options remain `undefined` and
   * disappear on `JSON.stringify()`. Per-component transform values from
   * {@link componentTransform} are intentionally not memoized and therefore
   * do not appear in the snapshot.
   *
   * The returned object aliases the internal cache; callers that need
   * isolation (e.g. {@link Avatar.toJSON}) clone it themselves.
   */
  resolved(): StyleOptions<D> {
    return this.#result as StyleOptions<D>;
  }

  /**
   * Returns the visibility probability (0–100) for the given component.
   * Aliases read the source component's user-set probability so a single
   * `<source>Probability` option propagates to every alias of the source.
   */
  #probability(component: Component): number {
    const raw = this.#options.componentProbability(component.sourceName());

    return raw ?? component.probability();
  }

  #isVisible(name: string, component: Component): boolean {
    return this.#prng.bool(`${name}Probability`, this.#probability(component));
  }

  /**
   * Resolves a named color to its final stop list, applying contrast sorting
   * and `notEqualTo` filtering from the style definition. Detects circular
   * references between colors and throws {@link CircularColorReferenceError}.
   */
  #resolveColor(name: string): readonly string[] {
    const userColors = this.#options.color(name);
    const styleColor = this.#style.colors().get(name);
    const source = userColors ?? styleColor?.values() ?? [];

    let candidates = source.map((c) => Color.toHex(c));
    const fill = this.colorFill(name);
    const stops = fill === 'solid' ? 1 : this.#colorFillStops(name);

    if (!styleColor) {
      return this.#prng.shuffle(`${name}Color`, candidates).slice(0, stops);
    }

    // Detect circular references (e.g. a.contrastTo = b, b.contrastTo = a)
    if (this.#colorResolving.includes(name)) {
      throw new CircularColorReferenceError(this.#colorResolving.concat(name));
    }

    this.#colorResolving.push(name);
    const contrastTo = styleColor.contrastTo();
    const notEqualTo = styleColor.notEqualTo();

    try {
      if (contrastTo) {
        const refColor = this.color(contrastTo)[0];

        if (refColor) {
          candidates = Color.sortByContrast(candidates, refColor);
        }
      }

      if (notEqualTo.length > 0) {
        const excluded: string[] = [];

        for (const ref of notEqualTo) {
          for (const color of this.color(ref)) {
            excluded.push(color);
          }
        }

        candidates = Color.filterNotEqualTo(candidates, excluded);
      }
    } finally {
      this.#colorResolving.pop();
    }

    // Skip shuffle when sorted by contrast to preserve the ordering
    const ordered = contrastTo
      ? candidates
      : this.#prng.shuffle(`${name}Color`, candidates);

    return ordered.slice(0, stops);
  }

  #colorFillStops(name: string): number {
    return (
      this.#prng.integer(`${name}ColorFillStops`, this.#options.colorFillStops(name)) ?? 2
    );
  }

  #pickFloat(
    name: string,
    suffix: string,
    range: readonly number[],
    fallback: number,
  ): number {
    return this.#prng.float(`${name}${suffix}`, range) ?? fallback;
  }

  #memo<T>(key: string, compute: () => T): T {
    if (key in this.#result) {
      return this.#result[key] as T;
    }

    const value = compute();

    this.#result[key] = value;

    return value;
  }
}

File: src/dart/core/lib/src/resolver.dart
/// Derives every deterministic value for an avatar from the style, the user
/// options, and a seeded PRNG, exposing them as memoized named accessors.
library;

import 'error/circular_color_reference_error.dart';
import 'options.dart';
import 'prng/prng.dart';
import 'range.dart';
import 'style.dart';
import 'style/color.dart';
import 'style/component.dart';
import 'utils/color.dart';

/// Bundles the three inputs needed to derive any deterministic value for an
/// avatar — the [Style], the validated user [Options], and a seeded [Prng] —
/// and exposes them as named accessors. Each accessor memoizes its result so
/// that repeated calls cannot drift. The memo also serves as the
/// informational snapshot returned by [resolved] — every value the resolver
/// picks during one resolution lands there, except for the raw seed.
class Resolver {
  final Style _style;
  final Options _options;
  final Prng _prng;
  final List<String> _colorResolving = [];

  // The memo doubles as the resolved-options snapshot: the first write wins,
  // and the map's insertion order is the first-resolution order the JSON
  // envelope emits (Dart maps preserve it natively; Go had to track the order
  // explicitly). Unset values are recorded as null and filtered on exposure.
  final Map<String, Object?> _result = {};

  Resolver(Style style, Options options)
      : _style = style,
        _options = options,
        _prng = Prng(options.seed() ?? '');

  /// Deliberately not memoized — the seed is the only input we keep out of
  /// the [resolved] snapshot, so a serialized avatar never leaks it.
  String seed() => _options.seed() ?? '';

  double? size() => _memo('size', () => _options.size());

  bool idRandomization() =>
      _memo('idRandomization', () => _options.idRandomization() ?? false);

  String? title() => _memo('title', () => _options.title());

  String flip() =>
      _memo('flip', () => _prng.pick('flip', _options.flip()) ?? 'none');

  String fontFamily() => _memo(
        'fontFamily',
        () => _prng.pick('fontFamily', _options.fontFamily()) ?? 'system-ui',
      );

  double fontWeight() => _memo(
        'fontWeight',
        () => _prng.pick('fontWeight', _options.fontWeight()) ?? 400.0,
      );

  double scale() => _memoFloat('scale', _options.scale(), 1);

  double borderRadius() =>
      _memoFloat('borderRadius', _options.borderRadius(), 0);

  double rotate() => _memoFloat('rotate', _options.rotate(), 0);

  double translateX() => _memoFloat('translateX', _options.translateX(), 0);

  double translateY() => _memoFloat('translateY', _options.translateY(), 0);

  /// Selects a variant for the given component, or `null` when the component
  /// is unknown or rolled invisible. Depending on what was passed as
  /// `${name}Variant` in the input data:
  ///
  /// - unset: the PRNG picks from all style variants using their weights.
  /// - otherwise: the PRNG picks using the user-supplied weighted map (a bare
  ///   string or string list is normalized to weight `1` each in
  ///   [Options.componentVariant]).
  ///
  /// Only variants that exist in the style definition are considered. User
  /// option lookup uses the component's source name (so a single option
  /// propagates to every alias), while the PRNG keys use the element's own
  /// name — each alias rolls its own visibility and variant.
  String? variant(String name) {
    return _memo('${name}Variant', () {
      final component = _style.components[name];

      if (component == null || !_isVisible(name, component)) {
        return null;
      }

      final raw = _options.componentVariant(component.sourceName);
      final variants = component.variants();
      final weights = <String, double>{};

      if (raw == null) {
        for (final entry in variants.entries) {
          weights[entry.key] = entry.value.weight();
        }
      } else {
        for (final entry in raw.entries) {
          if (variants.containsKey(entry.key)) {
            weights[entry.key] = entry.value;
          }
        }
      }

      return _prng.weightedPick('${name}Variant', weights);
    });
  }

  /// Resolves a named color to its final stop list.
  ///
  /// The memo is also a DoS guard: a color already resolved this pass is
  /// returned from the snapshot instead of being recomputed. Without it, a
  /// graph where each color references the next via both `contrastTo` and
  /// `notEqualTo` re-resolves exponentially (a schema-valid hang).
  List<String> color(String name) =>
      _memo('${name}Color', () => _resolveColor(name));

  String colorFill(String name) => _memo(
        '${name}ColorFill',
        () =>
            _prng.pick('${name}ColorFill', _options.colorFill(name)) ?? 'solid',
      );

  /// Memoized like every float option, but only ever called while a gradient
  /// def is built — `${name}ColorAngle` therefore appears in the snapshot
  /// only for colors that actually rendered as a gradient.
  double colorAngle(String name) =>
      _memoFloat('${name}ColorAngle', _options.colorAngle(name), 0);

  /// Picks the rotate/translateX/translateY/scale values for a single
  /// component. Memoized per `name`, so the four values land in [resolved] as
  /// `${name}Rotate` / `${name}TranslateX` / `${name}TranslateY` /
  /// `${name}Scale` for downstream introspection — recorded in exactly that
  /// order, which the snapshot key order depends on.
  ({double rotate, double translateX, double translateY, double scale})
      componentTransform(String name) {
    final component = _style.components[name];

    // Record fields evaluate in source order: Rotate, TranslateX,
    // TranslateY, Scale — the same order the JS port memoizes them in.
    return (
      rotate: _memoFloat('${name}Rotate', component?.rotate(), 0),
      translateX:
          _memoFloat('${name}TranslateX', component?.translate().x(), 0),
      translateY:
          _memoFloat('${name}TranslateY', component?.translate().y(), 0),
      scale: _memoFloat('${name}Scale', component?.scale(), 1),
    );
  }

  /// Returns an informational snapshot of every value the resolver picked.
  /// Includes top-level options (scale/rotate/translate/…), per-component
  /// variants/colors, and per-component transform picks. The raw seed is
  /// deliberately excluded; unset values are recorded as `null` and filtered
  /// out on exposure (Avatar drops them, like `JSON.stringify` drops
  /// `undefined` properties in JS).
  ///
  /// The returned map aliases the internal cache; callers that need isolation
  /// (e.g. `Avatar.toJson`) clone it themselves.
  Map<String, Object?> resolved() => _result;

  /// Returns the visibility probability (0–100) for the given component.
  /// Aliases read the source component's user-set probability so a single
  /// `<source>Probability` option propagates to every alias of the source.
  double _probability(Component component) =>
      _options.componentProbability(component.sourceName) ??
      component.probability();

  // `${name}Probability` is a PRNG key but never a memo key — visibility
  // rolls are not part of the resolved snapshot in any port.
  bool _isVisible(String name, Component component) =>
      _prng.boolean('${name}Probability', _probability(component));

  /// Resolves a named color to its final stop list, applying contrast sorting
  /// and `notEqualTo` filtering from the style definition. Detects circular
  /// references between colors and throws [CircularColorReferenceError].
  List<String> _resolveColor(String name) {
    final userColors = _options.color(name);
    final styleColor = _style.colors[name];
    final source = userColors ?? styleColor?.values() ?? const <String>[];

    var candidates = [for (final c in source) Color.toHex(c)];

    // colorFill is memoized inside this computation, so it lands in the
    // snapshot before `${name}Color` — the memo writes after compute returns.
    final fill = colorFill(name);
    final stops = fill == 'solid' ? 1 : _colorFillStops(name);

    if (styleColor == null) {
      return _takeN(_prng.shuffle('${name}Color', candidates), stops);
    }

    // Detect circular references (e.g. a.contrastTo = b, b.contrastTo = a).
    if (_colorResolving.contains(name)) {
      throw CircularColorReferenceError([..._colorResolving, name]);
    }

    _colorResolving.add(name);
    final contrastTo = _contrastTo(styleColor);
    final notEqualTo = styleColor.notEqualTo();

    try {
      if (contrastTo != null) {
        final refColors = color(contrastTo);

        if (refColors.isNotEmpty) {
          candidates = Color.sortByContrast(candidates, refColors[0]);
        }
      }

      if (notEqualTo.isNotEmpty) {
        final excluded = <String>[];

        for (final ref in notEqualTo) {
          excluded.addAll(color(ref));
        }

        candidates = Color.filterNotEqualTo(candidates, excluded);
      }
    } finally {
      _colorResolving.removeLast();
    }

    // Skip the shuffle when sorted by contrast, to preserve that ordering.
    final ordered = contrastTo != null
        ? candidates
        : _prng.shuffle('${name}Color', candidates);

    return _takeN(ordered, stops);
  }

  // The JS port's truthy check: an empty `contrastTo` counts as unset.
  static String? _contrastTo(ColorDefinition styleColor) {
    final value = styleColor.contrastTo();

    return value == null || value.isEmpty ? null : value;
  }

  /// Draws the gradient stop count, defaulting to `2`. Not memoized — the
  /// `${name}ColorFillStops` PRNG key never appears in the snapshot.
  int _colorFillStops(String name) {
    final range = _options.colorFillStops(name);

    return range != null ? _prng.integer('${name}ColorFillStops', range) : 2;
  }

  double _memoFloat(String key, Range? range, double fallback) =>
      _memo(key, () => range != null ? _prng.float(key, range) : fallback);

  T _memo<T>(String key, T Function() compute) {
    if (_result.containsKey(key)) {
      return _result[key] as T;
    }

    final value = compute();

    _result[key] = value;

    return value;
  }

  /// Returns a copy of the first [n] elements of [list] (or all of them when
  /// [n] is larger), mirroring the JS `slice(0, stops)`. Negative counts are
  /// clamped to 0, like the Go and Rust ports — the schema forbids them.
  static List<String> _takeN(List<String> list, int n) {
    if (n > list.length) {
      n = list.length;
    }

    if (n < 0) {
      n = 0;
    }

    return list.sublist(0, n);
  }
}
